analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs

```diff
@@ -0,0 +1,172 @@
+            {
+                Union(parents, dependency.From, dependency.To);
+            }
+
+            var disjointSets = DisjointSets(parents);
+            if (disjointSets.Count > 1)
+            {
+                var secondaryLocations = SecondaryLocations(symbol, disjointSets);
+                foreach (var primaryLocation in LocationIdentifiers<ClassDeclarationSyntax>(symbol))
+                {
+                    var diagnostic = Diagnostic.Create(Rule, primaryLocation, secondaryLocations.ToAdditionalLocations(), secondaryLocations.ToProperties(), disjointSets.Count);
+                    symbolEndContext.ReportIssue(CSharpFacade.Instance.GeneratedCodeRecognizer, diagnostic);
+                }
+            }
+        });
+    }
+
+    private static string BlockName(SyntaxNode block) =>
+        block switch
+        {
+            MethodDeclarationSyntax method => method.GetName(),
+            AccessorDeclarationSyntax { Parent.Parent: PropertyDeclarationSyntax property } => property.GetName(),
+            _ => null
+        };
+
+    private static void Union(IDictionary<string, string> parents, string from, string to) =>
+        parents[FindRoot(parents, from)] = FindRoot(parents, to);
+
+    private static string FindRoot(IDictionary<string, string> parents, string element) =>
+        parents[element] is var root && root != element ? FindRoot(parents, root) : root;
+
+    private static List<List<string>> DisjointSets(IDictionary<string, string> parents) =>
+        parents.GroupBy(x => FindRoot(parents, x.Key), x => x.Key).Select(x => x.OrderBy(x => x).ToList()).OrderBy(x => x[0]).ToList();
+
+    private static IEnumerable<Location> LocationIdentifiers<T>(ISymbol symbol) where T : SyntaxNode =>
+        symbol.DeclaringSyntaxReferences.Select(x => x.GetSyntax()).OfType<T>().Select(x => x.GetIdentifier()?.GetLocation());
+
+    private static ImmutableHashSet<string> RelevantMemberNames(INamedTypeSymbol symbol)
+    {
+        var builder = ImmutableHashSet.CreateBuilder<string>();
+        foreach (var member in symbol.GetMembers())
+        {
+            switch (member)
+            {
+                // Constructors are not considered because they have to be split anyway
+                // Accessors are not considered because they are part of properties, that are considered as a whole
+                case IMethodSymbol method when !method.IsConstructor() && !method.IsStaticConstructor() && method.AssociatedSymbol is not IPropertySymbol:
+                    builder.Add(method.Name);
+                    break;
+                // Primary constructor parameters may or may not generate fields, and must be considered
+                case IMethodSymbol method when method.IsPrimaryConstructor():
+                    builder.UnionWith(method.Parameters.Where(IsService).Select(x => x.Name));
+                    break;
+                case IFieldSymbol field when IsService(field) && !field.IsImplicitlyDeclared:
```

Comment:
`ImplicitlyDeclared` is pulling double-duty here, right?

- Property backers
- Primary constructor parameter backers

I would add a comment to make this a bit more clear:

```csharp
// Exclude backing fields for auto-properties and primary constructor parameters
```